### PR TITLE
[#25] [KMM] [Backend] As a user, I can log out

### LINKED ISSUE #25: [KMM] [Backend] As a user, I can log out
## Why

When user tap logout, the app needs to call API to revoke token and remove the token from storage.

## Acceptance Criteria?

- Call endpoint:
`/api/v1/oauth/revoke`
```
{
    "token": "{{user_access_token}}",
    "client_id": "{{client_id}}",
    "client_secret": "{{client_secret}}"
}
```

## Resources

[Postman - Nimble Survey API](https://documenter.getpostman.com/view/11835486/T1LJjTqj#bc9cd5d4-a4c8-4692-bfaf-bdc73f080588)

### PR DESCRIPTION
**Pull request by blyscuit**
- close #25 

## What happened

Add remove token from storage when user logout.

## Insight

- Revoke API is not callable because the client type is not supported.
- Setting saves object with key from object's properties. So we have to [remove key recursively](https://github.com/russhwolf/multiplatform-settings/issues/136).

Screen Shot 2022-12-22 at 12 24 02

 
## Proof Of Work

Unit test passes.

### CODE REVIEW

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/data/database/core/DataStore.kt [hunk 1]
```diff
@@ -38,4 +39,11 @@ class DataStoreImpl(private val settings: Settings): DataStore {
     override fun <T> save(serializer: KSerializer<T>, key: String, value: T) {
         settings.encodeValue(serializer, key, value)
     }
+
+    override fun remove(key: String) {
+        settings
+            .keys
+            .filter { it.startsWith(key) }
+            .map { settings.remove(it) }
+    }
```
Comment (minhnimble) [inline shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/data/database/core/DataStore.kt]: Just curious, why don't we do exact string comparison here? with the `startsWith` function, we might remove more keys unexpectedly if multiple keys have the same prefixes and if this is intentional, I think we should reflect that into the function name also, for example:
```suggestion
    override fun removeStartingWith(key: String) {
        settings
            .keys
            .filter { it.startsWith(key) }
            .map { settings.remove(it) }
    }
```
 What do you think? 🤔

File: shared/src/commonTest/kotlin/co/nimblehq/blisskmmic/data/datasource/LocalDataSourceTest.kt [hunk 2]
```diff
@@ -67,4 +67,19 @@ class LocalDataSourceTest: TestsWithMocks() {
                 it shouldBe tokenDB
             }
     }
+
+    @Test
+    fun `When removing token- it remove the key from the dataStore`() = runTest {
```
Comment (minhnimble) [inline shared/src/commonTest/kotlin/co/nimblehq/blisskmmic/data/datasource/LocalDataSourceTest.kt]: This test miss a front space I think, and there are several instances in this PR, please check and update them all 🙏
```suggestion
    fun `When removing token - it remove the key from the dataStore`() = runTest {
```

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/UITestLogout.kt [hunk 3]
```diff
@@ -1,15 +1,13 @@
 package co.nimblehq.blisskmmic.helpers.extensions.ios
 
-import co.nimblehq.blisskmmic.data.database.datasource.LocalDataSourceImpl
-import com.russhwolf.settings.Settings
+import co.nimblehq.blisskmmic.data.database.datasource.LocalDataSource
 import org.koin.core.component.KoinComponent
 import org.koin.core.component.inject
 
 object UITestLogout: KoinComponent {
```
Comment (minhnimble) [inline shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/UITestLogout.kt:7]: Just curious, why this file `UITestLogout.kt` is a kotlin file but it is under the `ios` extension? Shouldn't it be in `android` extension instead? or if it is used for iOS only, can we write it in Swift? 🤔

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/UITestLogout.kt [hunk 4]
```diff
@@ -1,15 +1,13 @@
...
 
     fun logOut() {
-        // TODO: Use logout use case to clear
-        val settings : Settings by inject()
-        settings.clear()
+        val localDataSource: LocalDataSource by inject()
```
Comment (minhnimble) [inline shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/UITestLogout.kt]: Even though we can inject the `localDataSource` directly inside the function, it is kinda a weird practice and it looks tightly-coupled here. Should we turn it into a seperated private val instead? 🤔
```
private val localDataSource: LocalDataSource by inject()

fun logOut() {
   localDataSource.removeToken()
}
```

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/data/database/core/DataStore.kt (hunk 1 above)
Comment (blyscuit) [reply shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/data/database/core/DataStore.kt]: The keys are mapped internally by the lib, this feature is in development from the lib so we have to add workaround. I will rename this to `removeObject(key:)`

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/UITestLogout.kt (hunk 3 above)
Comment (blyscuit) [reply shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/UITestLogout.kt:7]: This is a helper object for iOS to use, hence there's iOS in the package. It is under `common`.

**Review by doannimble — APPROVED** [review]
LGTM 🏃‍♂️

**Review by minhnimble — COMMENTED** [review]
LGTM 🚀

**Review by minhnimble — APPROVED** [review]

**Pull request merged**